sql/expression/function/strcmp.go

```diff
@@ -92,11 +88,21 @@ func (s *StrCmp) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 		return nil, nil
 	}
 
-	str2, err := sql.ConvertToString(expr2, sql.LongText)
-	if err != nil {
-		return nil, err
+	if sql.IsText(s.Left.Type()) && sql.IsText(s.Right.Type()) {
```

Comment:
You're on the right track here! Collation comparisons have to take coercibility into account. You can see an example of this here:
https://github.com/dolthub/go-mysql-server/blob/e28aad108618a29d0c60138a7974f932dfdf7706/sql/expression/comparison.go#L127-L137
And this is the MySQL documentation going over coercibility:
https://dev.mysql.com/doc/refman/8.0/en/charset-collation-coercibility.html
We don't fully support coercibility just yet, but it's something I'm going to be working on soon. In the meantime, our approximation is the best we have.